### PR TITLE
Add named profiles for multi-identity support

### PR DESCRIPTION
**Pull request by jeremy**
## Summary

- Replace the unreleased host infrastructure with named profiles that bundle identity (credential key) with environment (base_url, account_id, scope, etc.)
- Allow multiple Basecamp users to authenticate against the same instance with separate credentials stored under `profile:<name>` keys
- Add `bcq profile` commands for managing profiles (list, show, create, delete, set-default)

## Design

A profile is a named configuration bundle:

```json
{
  "default_profile": "personal",
  "profiles": {
    "personal": {
      "base_url": "https://3.basecampapi.com",
      "account_id": "12345"
    },
    "triage-bot": {
      "base_url": "https://3.basecampapi.com",
      "account_id": "12345",
      "scope": "full"
    }
  }
}
```

Two profiles can share a `base_url` but authenticate as different users. Credentials are stored under `profile:<name>` instead of the origin URL.

**Profile resolution order:** `--profile`/`-P` flag → `BCQ_PROFILE` env → `default_profile` config → single profile auto-use → interactive picker (if TTY) → no profile (origin-based, backwards compatible)

**Precedence:** flags > env vars > profile values > file configs > defaults

**No-profile mode:** When no profiles are configured, everything works as before with origin-based credentials.

## Commits

1. **Replace host infrastructure with named profiles** — config types, root.go, completion, TUI resolve, appctx
2. **Key credentials by profile name** — `credentialKey()`/`CredentialKey()` in auth.Manager
3. **Make commands profile-aware** — auth login/status/token, config set, doctor
4. **Add profile management commands** — `bcq profile list/show/create/delete/set-default`

## Test plan

- [x] `bcq profile create triage-bot` → prompts base_url/scope → starts OAuth login
- [x] `bcq --profile triage-bot auth status` shows profile name and auth state
- [x] Two profiles at same base_url maintain independent credentials
- [x] `bcq profile set-default triage-bot` → `bcq auth status` uses triage-bot
- [x] Config with no profiles works as before (origin-based credentials)
- [x] `bcq --profile <TAB>` completes profile names with base_url descriptions
- [x] `BCQ_PROFILE=dev bcq auth status` selects correct profile via env var
- [x] `bcq -P production --account 67890` flag overrides profile value (source: flag)
- [x] Profile-scoped cache created at `~/.cache/bcq/profiles/<name>/`

### CODE REVIEW

File: internal/cli/root.go [hunk 1]
```diff
@@ -222,26 +221,100 @@ func Execute() {
...
+func resolveProfile(cfg *config.Config, flags appctx.GlobalFlags) string {
+	// 1. --profile flag
+	if flags.Profile != "" {
+		return flags.Profile
+	}
+
+	// 2. BCQ_PROFILE env var
+	if profile := os.Getenv("BCQ_PROFILE"); profile != "" {
+		return profile
+	}
```
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/cli/root.go:257]: Fixed — `resolveProfile` now checks `len(cfg.Profiles)` before honoring both `--profile` flag and `BCQ_PROFILE` env var. If the named profile does not exist, it returns a clear error listing available profiles. If no profiles are configured at all, both are ignored and we fall through to no-profile mode.

File: internal/cli/root.go [hunk 2]
```diff
@@ -222,26 +221,100 @@ func Execute() {
...
+	if isInteractiveTTY(flags) {
+		if name, err := promptForProfile(cfg); err == nil {
+			return name
 		}
 	}
 
-	// Not a configured host name - treat as hostname/URL
-	return hostutil.Normalize(host)
+	// 6. No profile resolved
+	return ""
```
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/cli/root.go]: By design — when no profile is resolved, commands fall back to top-level config values (origin-based credentials). Commands that require an account/project will error at the resolution stage with actionable messages. Adding an error here would break the no-profile compatibility path.

File: internal/commands/profile.go [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+			}
+			if profileCfg.AccountID != "" {
+				profileEntry["account_id"] = profileCfg.AccountID
+			}
+			profilesMap[name] = profileEntry
+			configData["profiles"] = profilesMap
+
+			// If this is the first profile, set it as default
+			if len(profilesMap) == 1 {
+				configData["default_profile"] = name
```
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/commands/profile.go:266]: Fixed — now prints `Set "name" as default profile (first profile created)` when auto-setting the default.

File: internal/cli/root.go [hunk 4]
```diff
@@ -222,26 +221,100 @@ func Execute() {
...
+// 5. Multiple profiles → interactive picker (if TTY)
+// 6. No profiles → empty string (use top-level config values)
+func resolveProfile(cfg *config.Config, flags appctx.GlobalFlags) string {
+	// 1. --profile flag
+	if flags.Profile != "" {
+		return flags.Profile
+	}
+
+	// 2. BCQ_PROFILE env var
+	if profile := os.Getenv("BCQ_PROFILE"); profile != "" {
```
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/cli/root.go:250]: Fixed — `resolveProfile` now validates both `--profile` flag and `BCQ_PROFILE` env against `cfg.Profiles` and returns a clear error with available profile names. Changed signature to return `(string, error)`.

File: internal/config/config.go [hunk 5]
```diff
@@ -277,6 +294,46 @@ func applyOverrides(cfg *Config, o FlagOverrides) {
...
+		cfg.TodolistID = p.TodolistID
+		cfg.Sources["todolist_id"] = "profile"
+	}
+	if p.Scope != "" {
+		cfg.Scope = p.Scope
+		cfg.Sources["scope"] = "profile"
+	}
+
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/config/config.go:342]: By design — the two-pass approach is documented in the function comment. Profile values are intentionally applied unconditionally, then `LoadFromEnv` and `ApplyOverrides` are re-applied afterward to restore env/flag precedence. This matches the stated precedence chain: flags > env > profile > file config > defaults.

File: internal/commands/profile.go [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+				return output.ErrUsage(fmt.Sprintf("Profile %q not found", name))
+			}
+			if _, ok := app.Config.Profiles[name]; !ok {
+				return output.ErrUsage(fmt.Sprintf("Profile %q not found", name))
+			}
+
+			// Remove credentials
+			credKey := "profile:" + name
+			store := app.Auth.GetStore()
+			_ = store.Delete(credKey) // Best-effort credential cleanup
```
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/commands/profile.go]: Fixed — credential deletion failure now prints a warning to stderr.

File: internal/commands/profile.go [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+			}
+
+			name := args[0]
+
+			// Check if profile already exists
+			if app.Config.Profiles != nil {
+				if _, exists := app.Config.Profiles[name]; exists {
+					return output.ErrUsage(fmt.Sprintf("Profile %q already exists", name))
+				}
+			}
```
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/commands/profile.go:210]: Fixed — added `isValidProfileName()` that rejects names with path separators, colons, or other special characters. Only letters, numbers, hyphens, and underscores are allowed, must start with alphanumeric.

File: internal/cli/root.go [hunk 8]
```diff
@@ -222,26 +221,100 @@ func Execute() {
...
-		for name, hostConfig := range cfg.Hosts {
-			if strings.ToLower(name) == hostLower {
-				return hostConfig.BaseURL
-			}
+	// 3. default_profile in config
+	if cfg.DefaultProfile != "" {
+		if _, ok := cfg.Profiles[cfg.DefaultProfile]; ok {
+			return cfg.DefaultProfile
+		}
+	}
```
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/cli/root.go:270]: Intentional — silently falling through to the next resolution step (single-profile auto-use, interactive picker) is the most user-friendly behavior. The stale default_profile gets corrected naturally when the user picks a new profile or creates a new one. A warning here would be noisy on every command invocation for a benign misconfiguration.

File: internal/completion/cache.go [hunk 9]
```diff
@@ -352,18 +351,18 @@ func loadConfigForCompletion() *configForCompletion {
...
+		loadProfilesFromFile(cfg, globalPath)
 	}
 
 	// Try local config
 	localPath := filepath.Join(".basecamp", "config.json")
-	loadHostsFromFile(cfg, localPath)
+	loadProfilesFromFile(cfg, localPath)
 
 	return cfg
 }
```
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/completion/cache.go:383]: Matches the existing pattern — completion has always loaded only global and local configs for performance (it runs on every tab press). System and repo configs are uncommon for profile definitions. The main config loader handles all layers at runtime.

File: internal/tui/resolve/persist.go [hunk 10]
```diff
@@ -126,15 +126,7 @@ func PromptAndPersistTodolistID(todolistID string) (bool, error) {
...
-// PromptAndPersistDefaultHost prompts the user to save a default host.
-func PromptAndPersistDefaultHost(hostName string) (bool, error) {
-	return PromptAndPersist(PersistOption{
-		Key:   "default_host",
-		Value: hostName,
-	})
+// PersistDefaultProfile is a convenience function for persisting a default profile.
+func PersistDefaultProfile(profileName, scope string) error {
+	return PersistValue("default_profile", profileName, scope)
 }
```
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/tui/resolve/persist.go:132]: Follows the same pattern as `PersistAccountID`, `PersistProjectID`, and `PersistTodolistID` — convenience functions available for future interactive flows. `PromptAndPersistDefaultProfile` is not needed because profile selection already has its own flow via `bcq profile create` and `bcq profile set-default`.

File: internal/commands/profile.go [hunk 11]
```diff
@@ -0,0 +1,431 @@
...
+	cmd.AddCommand(
+		newProfileListCmd(),
+		newProfileShowCmd(),
+		newProfileCreateCmd(),
+		newProfileDeleteCmd(),
+		newProfileSetDefaultCmd(),
+	)
+
+	return cmd
+}
```
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/commands/profile.go:46]: Acknowledged — will add profile command tests in a follow-up. The profile commands were manually tested against the full test plan in the PR description (all 9 items passing).

File: internal/commands/profile.go [hunk 12]
```diff
@@ -0,0 +1,431 @@
...
+
+			// Write config
+			data, err := json.MarshalIndent(configData, "", "  ")
+			if err != nil {
+				return fmt.Errorf("failed to marshal config: %w", err)
+			}
+			if err := os.WriteFile(configPath, append(data, '\n'), 0600); err != nil {
+				return fmt.Errorf("failed to write config: %w", err)
+			}
+
```
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/commands/profile.go]: Not a concern for a single-user CLI tool — `bcq profile create` is an interactive command run by one user at a time. The existing config commands (`config set`, `config project`) use the same read-modify-write pattern. Atomic writes would add complexity without practical benefit.

File: internal/cli/root.go [hunk 13]
```diff
@@ -48,22 +48,22 @@ func NewRootCmd() *cobra.Command {
...
+				// Re-apply env and flag overrides (they take precedence over profile values)
+				config.LoadFromEnv(cfg)
+				config.ApplyOverrides(cfg, config.FlagOverrides{
+					Account:  flags.Account,
+					Project:  flags.Project,
+					Todolist: flags.Todolist,
+					CacheDir: flags.CacheDir,
+				})
+				// Profile-scoped cache
+				cfg.CacheDir = filepath.Join(cfg.CacheDir, "profiles", profileName)
```
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/cli/root.go]: Fixed — profile-scoped cache directory is now only applied when `--cache-dir` was not explicitly set via flag.

File: internal/cli/root.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/cli/root.go]: Updated — `resolveProfile` now returns an error when multiple profiles exist in non-interactive mode: `"multiple profiles configured but none selected; use --profile or set default_profile"`

File: internal/config/config.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/config/config.go:342]: Updated — improved the ApplyProfile comment with a clear two-pass explanation documenting the precedence chain.

File: internal/cli/root.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/cli/root.go:270]: Updated — now prints a warning to stderr when default_profile references a non-existent profile, then falls through to the next resolution step.

File: internal/completion/cache.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/completion/cache.go:383]: Updated — `loadConfigForCompletion` now loads all 4 config layers: system (`/etc/basecamp/config.json`), global, repo (walks up to .git), and local.

File: internal/tui/resolve/persist.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/tui/resolve/persist.go:132]: Updated — added `PromptAndPersistDefaultProfile` to match the pattern of the other convenience functions.

File: internal/commands/profile.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/commands/profile.go:46]: Done — added `profile_test.go` with 37 tests covering command structure, name validation, show/list/delete/set-default commands, config file integration, and edge cases.

File: internal/commands/profile.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/commands/profile.go]: Updated — added `atomicWriteJSON` helper using temp file + os.Rename. All three config write sites in profile.go now use it.

File: internal/commands/config.go [hunk 14]
```diff
@@ -170,14 +170,15 @@ Valid keys: account_id, project_id, todolist_id, base_url, cache_dir, cache_enab
...
+				"account_id":      true,
+				"project_id":      true,
+				"todolist_id":     true,
+				"base_url":        true,
+				"cache_dir":       true,
+				"cache_enabled":   true,
+				"format":          true,
+				"scope":           true,
+				"default_profile": true,
 			}
```
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/commands/config.go:182]: Fixed — `config set default_profile` now validates the profile exists in the config file before writing. Returns an error with available profile names if the profile is not found.

File: internal/completion/cache.go [hunk 15]
```diff
@@ -292,58 +292,61 @@ func (s *Store) Accounts() []CachedAccount {
...
 type configForCompletion struct {
-	Hosts map[string]*hostConfig `json:"hosts,omitempty"`
+	Profiles map[string]*profileConfig `json:"profiles,omitempty"`
 }
 
-// loadConfigForCompletion loads config files to get hosts for completion.
-// This is a simplified version that only reads what's needed for host completion.
+// loadConfigForCompletion loads config files from all layers to get profiles
+// for completion. Layers are loaded in precedence order (system → global →
+// repo → local) so that higher-precedence layers override lower ones.
```
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/completion/cache.go:340]: Fixed — comment now reads "loaded from lowest to highest precedence ... so that later layers override earlier ones."

File: internal/cli/root.go [hunk 16]
```diff
@@ -48,21 +49,26 @@ func NewRootCmd() *cobra.Command {
...
+				// Re-apply env and flag overrides (they take precedence over profile values)
+				config.LoadFromEnv(cfg)
+				config.ApplyOverrides(cfg, config.FlagOverrides{
+					Account:  flags.Account,
+					Project:  flags.Project,
+					Todolist: flags.Todolist,
+					CacheDir: flags.CacheDir,
+				})
+				// Profile-scoped cache (only if cache dir was not explicitly set via flag)
+				if flags.CacheDir == "" {
```
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/cli/root.go:70]: Fixed — profile-scoped cache directory now also checks `BCQ_CACHE_DIR` env var. Only applies profile scoping when neither the flag nor the env var is set.

File: internal/cli/root.go [hunk 17]
```diff
@@ -222,26 +227,121 @@ func Execute() {
...
-			if strings.ToLower(name) == hostLower {
-				return hostConfig.BaseURL
+	// 2. BCQ_PROFILE env var
+	if profile := os.Getenv("BCQ_PROFILE"); profile != "" {
+		if len(cfg.Profiles) > 0 {
+			if _, ok := cfg.Profiles[profile]; !ok {
+				return "", fmt.Errorf("unknown profile %q from BCQ_PROFILE (available: %s)", profile, profileNames(cfg))
 			}
 		}
+		return profile, nil
```
[the remark this answers is not in the document]
Comment (jeremy) [reply internal/cli/root.go:256]: Fixed — `resolveProfile` now returns a clear error when `--profile` or `BCQ_PROFILE` is specified but no profiles are configured, with guidance to run `bcq profile create`.
